In [1]:
print("hello world")

hello world


In [2]:
import pandas as pd

In [3]:
df = pd.read_pickle('data/02_indicators.pkl')
df

Price                   close        high         low        open  \
date       ticker                                                   
2018-08-15 A        60.991581   62.555705   60.511034   62.131698   
           AAPL     49.666737   49.784856   49.215521   49.425773   
           ABBV     69.514221   69.756757   68.201693   68.422824   
           ABNB           NaN         NaN         NaN         NaN   
           ABT      54.906963   54.984821   54.560928   54.889661   
...                       ...         ...         ...         ...   
2026-08-12 XYZ      78.279999   79.199997   77.230003   78.580002   
           YUM     150.339996  150.389999  144.429993  144.610001   
           ZBH      97.889999   97.940002   94.279999   95.690002   
           ZBRA    378.369995  386.230011  374.019989  384.230011   
           ZTS      73.540001   74.389999   71.989998   74.389999   

Price                   volume  garman_klass_vol        rsi    bb_low  \
date       ticker                                                       
2018-08-15 A         6071500.0          0.000420        NaN       NaN   
           AAPL    115230400.0          0.000057        NaN       NaN   
           ABBV      6861500.0          0.000157        NaN       NaN   
           ABNB            NaN               NaN        NaN       NaN   
           ABT       3724900.0          0.000030        NaN       NaN   
...                        ...               ...        ...       ...   
2026-08-12 XYZ       4343100.0          0.000312  49.935831  4.342263   
           YUM       3080700.0          0.000234  47.965239  4.974626   
           ZBH       2733600.0          0.000526  59.070244  4.485455   
           ZBRA       748500.0          0.000425  78.001656  5.402393   
           ZTS       7186400.0          0.000487  43.338929  4.303346   

Price                bb_mid   bb_high       atr      macd  dollar_volume  
date       ticker                                                         
2018-08-15 A            NaN       NaN       NaN       NaN     370.310384  
           AAPL         NaN       NaN       NaN       NaN    5723.117925  
           ABBV         NaN       NaN       NaN       NaN     476.971829  
           ABNB         NaN       NaN       NaN       NaN            NaN  
           ABT          NaN       NaN       NaN       NaN     204.522948  
...                     ...       ...       ...       ...            ...  
2026-08-12 XYZ     4.398908  4.455553 -0.668036  0.144249     339.977863  
           YUM     5.014113  5.053600  3.248934 -1.234277     463.152427  
           ZBH     4.555733  4.626010  0.354879  0.997090     267.592102  
           ZBRA    5.721512  6.040631  1.556197  3.605673     283.209941  
           ZTS     4.340344  4.377342 -0.398774 -0.195375     528.487863  

[1010024 rows x 13 columns]

In [4]:
print(df.shape)
df.head()

(1010024, 13)


Price                  close       high        low       open       volume  \
date       ticker                                                            
2018-08-15 A       60.991581  62.555705  60.511034  62.131698    6071500.0   
           AAPL    49.666737  49.784856  49.215521  49.425773  115230400.0   
           ABBV    69.514221  69.756757  68.201693  68.422824    6861500.0   
           ABNB          NaN        NaN        NaN        NaN          NaN   
           ABT     54.906963  54.984821  54.560928  54.889661    3724900.0   

Price              garman_klass_vol  rsi  bb_low  bb_mid  bb_high  atr  macd  \
date       ticker                                                              
2018-08-15 A               0.000420  NaN     NaN     NaN      NaN  NaN   NaN   
           AAPL            0.000057  NaN     NaN     NaN      NaN  NaN   NaN   
           ABBV            0.000157  NaN     NaN     NaN      NaN  NaN   NaN   
           ABNB                 NaN  NaN     NaN     NaN      NaN  NaN   NaN   
           ABT             0.000030  NaN     NaN     NaN      NaN  NaN   NaN   

Price              dollar_volume  
date       ticker                 
2018-08-15 A          370.310384  
           AAPL      5723.117925  
           ABBV       476.971829  
           ABNB              NaN  
           ABT        204.522948

In [5]:
df.unstack('ticker')['dollar_volume'].resample('ME').mean().stack('ticker').to_frame('dollar_volume')

dollar_volume
date       ticker               
2018-08-31 A          153.014041
           AAPL      5854.973599
           ABBV       294.747135
           ABNB              NaN
           ABT        248.662236
...                          ...
2026-08-31 XYZ        476.457452
           YUM        370.952232
           ZBH        306.447240
           ZBRA       389.639119
           ZTS        699.134300

[48791 rows x 1 columns]

In [6]:
df.unstack('ticker')['dollar_volume']

ticker,A,AAPL,ABBV,ABNB,ABT,ACGL,ACN,ADBE,ADI,ADM,...,WY,WYNN,XEL,XOM,XYL,XYZ,YUM,ZBH,ZBRA,ZTS
date,,,,,,,,,,,,,,,,,,,,,
2018-08-15,370.310384,5723.117925,476.971829,NaN,204.522948,18.749262,235.219392,687.602256,164.156614,176.441618,...,76.485455,492.179679,122.595925,879.512008,58.019478,942.796988,109.424819,96.494756,67.242197,170.212504
2018-08-16,194.378001,5745.037624,443.705087,NaN,206.191376,22.428435,280.005503,602.413355,149.233079,241.850968,...,149.216967,327.585875,151.197557,629.850612,55.175890,594.827569,125.257699,79.360711,51.659099,209.346401
2018-08-17,149.219022,7283.894125,444.566002,NaN,266.963523,21.340635,393.590997,674.249932,234.955527,240.234869,...,125.620544,235.318571,131.878205,558.359191,164.246053,405.426068,120.997204,91.368167,49.872495,230.432992
2018-08-20,98.403401,6166.563427,429.869002,NaN,283.811867,15.452419,186.489441,724.628672,175.377914,157.974099,...,79.415730,330.970146,90.280948,450.344875,64.232676,525.515757,126.901142,81.996321,50.445168,100.648883
2018-08-21,182.487728,5315.743130,246.003902,NaN,226.605763,22.628829,186.752124,545.478350,336.433092,115.732693,...,50.463986,243.072835,124.577168,507.931747,56.900180,645.716703,97.964407,97.837851,57.588082,104.378418
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-08-06,164.010932,14402.145672,1114.534652,916.072400,1058.137543,118.601906,890.798663,1223.284098,1351.260294,337.853302,...,191.450176,137.571257,414.083213,1827.526378,141.337507,921.183513,363.812145,291.395065,267.497849,1379.624882
2026-08-07,222.633446,10780.909748,1213.149395,2833.984167,712.257530,110.809700,646.298164,1260.437005,1106.426354,234.675800,...,188.959459,153.679878,396.829808,2806.213186,182.525440,474.158000,268.594006,211.125892,276.230706,928.929083
2026-08-10,260.542570,13813.901688,1043.829720,1844.395700,608.619599,124.256775,708.490275,1106.197661,1075.157551,440.702808,...,162.994056,120.143373,334.046284,3137.391009,171.144521,332.292568,303.201983,320.571726,213.388344,719.116318


In [7]:
df.unstack('ticker')

Price            close                                                  \
ticker               A        AAPL        ABBV        ABNB         ABT   
date                                                                     
2018-08-15   60.991581   49.666737   69.514221         NaN   54.906963   
2018-08-16   61.717098   50.394360   70.063492         NaN   55.503883   
2018-08-17   60.925617   51.400726   70.484360         NaN   55.988323   
2018-08-20   60.746590   50.899899   69.642609         NaN   55.919113   
2018-08-21   61.180008   50.800686   69.721092         NaN   56.014278   
...                ...         ...         ...         ...         ...   
2026-08-06  141.339996  312.140808  243.869995  151.639999  107.959999   
2026-08-07  145.970001  313.059998  246.039993  178.070007  107.809998   
2026-08-10  148.720001  308.260010  247.970001  184.699997  108.620003   
2026-08-11  149.490005  304.910004  250.089996  184.979996  109.720001   
2026-08-12  148.240005  302.250000  248.759995  180.095001  110.910004   

Price                                                                 ...  \
ticker           ACGL         ACN        ADBE         ADI        ADM  ...   
date                                                                  ...   
2018-08-15  28.774189  140.900558  250.520004   81.173225  39.090240  ...   
2018-08-16  28.954861  143.078949  248.889999   81.606102  39.651600  ...   
2018-08-17  29.126020  144.835693  245.699997   81.883156  39.967869  ...   
2018-08-20  29.040442  145.037674  249.759995   81.155907  40.141815  ...   
2018-08-21  29.221111  144.589752  251.500000   83.043243  40.118099  ...   
...               ...         ...         ...         ...        ...  ...   
2026-08-06  99.290001  171.110001  260.239990  377.309998  77.021156  ...   
2026-08-07  98.480003  175.720001  265.209991  389.929993  76.106956  ...   
2026-08-10  98.250000  178.250000  272.959991  383.929993  79.982361  ...   
2026-08-11  98.260002  179.820007  263.709991  385.299988  79.952553  ...   
2026-08-12  97.290001  180.139999  258.750000  384.429993  79.227158  ...   

Price      dollar_volume                                                   \
ticker                WY        WYNN         XEL          XOM         XYL   
date                                                                        
2018-08-15     76.485455  492.179679  122.595925   879.512008   58.019478   
2018-08-16    149.216967  327.585875  151.197557   629.850612   55.175890   
2018-08-17    125.620544  235.318571  131.878205   558.359191  164.246053   
2018-08-20     79.415730  330.970146   90.280948   450.344875   64.232676   
2018-08-21     50.463986  243.072835  124.577168   507.931747   56.900180   
...                  ...         ...         ...          ...         ...   
2026-08-06    191.450176  137.571257  414.083213  1827.526378  141.337507   
2026-08-07    188.959459  153.679878  396.829808  2806.213186  182.525440   
2026-08-10    162.994056  120.143373  334.046284  3137.391009  171.144521   
2026-08-11    120.401246  107.900086  288.087550  2182.240749  175.088052   
2026-08-12    107.431165  188.735039  268.623355  2098.228708  194.588087   

Price                                                                    
ticker             XYZ         YUM         ZBH        ZBRA          ZTS  
date                                                                     
2018-08-15  942.796988  109.424819   96.494756   67.242197   170.212504  
2018-08-16  594.827569  125.257699   79.360711   51.659099   209.346401  
2018-08-17  405.426068  120.997204   91.368167   49.872495   230.432992  
2018-08-20  525.515757  126.901142   81.996321   50.445168   100.648883  
2018-08-21  645.716703   97.964407   97.837851   57.588082   104.378418  
...                ...         ...         ...         ...          ...  
2026-08-06  921.183513  363.812145  291.395065  267.497849  1379.624882  
2026-08-07  474.158000  268.594006  211.125892  276.230706   928.

In [8]:
df.unstack('ticker')['dollar_volume'].resample('ME').mean()

ticker,A,AAPL,ABBV,ABNB,ABT,ACGL,ACN,ADBE,ADI,ADM,...,WY,WYNN,XEL,XOM,XYL,XYZ,YUM,ZBH,ZBRA,ZTS
date,,,,,,,,,,,,,,,,,,,,,
2018-08-31,153.014041,5854.973599,294.747135,NaN,248.662236,21.493240,245.513273,683.831069,249.150321,136.241891,...,77.524814,285.093860,121.171649,515.741295,61.366016,848.732970,112.687767,106.767838,52.061453,160.059864
2018-09-30,123.257081,7483.296675,346.325562,NaN,344.906256,33.195092,322.641604,849.769324,272.947930,119.569277,...,91.033384,359.163483,133.277889,700.481940,64.043085,1241.184916,125.538310,125.658699,55.846486,171.697576
2018-10-31,176.723221,7139.701668,375.788405,NaN,478.142171,27.136687,332.837455,1163.944148,299.348712,134.238429,...,115.469611,282.922345,132.865101,760.328145,80.912903,1582.894439,153.519280,138.478795,77.547178,199.400395
2018-11-30,187.605549,8333.096335,485.583011,NaN,392.594305,28.335316,324.441680,920.189366,253.254916,141.734281,...,86.996433,355.582169,217.376932,765.462942,68.818237,1223.181116,178.192640,115.918710,84.510682,208.652350
2018-12-31,172.944904,7319.299544,418.578258,NaN,463.053114,32.324502,408.404716,1167.567529,213.441554,144.001852,...,138.181374,314.208579,213.792334,1029.906300,66.983520,1096.528512,215.135557,161.692742,80.346548,185.721949
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-04-30,212.683888,11433.614858,1394.386418,438.839102,1190.327935,200.588254,955.879566,1273.139998,1169.567012,227.818027,...,106.186182,139.359100,357.710538,3045.493481,313.615902,333.124385,250.701627,229.012970,175.222064,380.488173
2026-05-31,312.257707,14562.781103,1158.987803,563.194936,1080.778751,246.003800,1029.567463,1170.459676,1864.762579,332.057695,...,129.280464,156.121431,448.021813,2649.690346,220.429735,452.449377,287.640510,188.879403,308.030033,799.470692
2026-06-30,313.490334,19562.997080,2035.292437,557.963874,1252.316612,217.615333,1539.782603,1892.044633,2340.480108,284.722074,...,146.661560,160.479344,505.597671,2577.383586,220.503832,463.372034,358.197746,198.361409,228.071479,572.913364


In [9]:
last_cols = [c for c in df.columns.unique(0) if c not in ['dollar_volume','volume','open','low',]]
last_cols

['close',
 'high',
 'garman_klass_vol',
 'rsi',
 'bb_low',
 'bb_mid',
 'bb_high',
 'atr',
 'macd']

In [10]:
df.unstack()[last_cols].resample('ME').last().stack('ticker')

Price                   close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2018-08-31 A        63.639286   63.648711          0.000014  94.747320   
           AAPL     53.774918   54.067851          0.000070  96.803688   
           ABBV     68.465599   68.972064          0.000048  75.028397   
           ABNB           NaN         NaN               NaN        NaN   
           ABT      57.822296   58.168332          0.000046  96.406869   
...                       ...         ...               ...        ...   
2026-08-31 XYZ      78.279999   79.199997          0.000312  49.935831   
           YUM     150.339996  150.389999          0.000234  47.965239   
           ZBH      97.889999   97.940002          0.000526  59.070244   
           ZBRA    378.369995  386.230011          0.000425  78.001656   
           ZTS      73.540001   74.389999          0.000487  43.338929   

Price                bb_low    bb_mid   bb_high       atr      macd  
date       ticker                                                    
2018-08-31 A            NaN       NaN       NaN       NaN       NaN  
           AAPL         NaN       NaN       NaN       NaN       NaN  
           ABBV         NaN       NaN       NaN       NaN       NaN  
           ABNB         NaN       NaN       NaN       NaN       NaN  
           ABT          NaN       NaN       NaN       NaN       NaN  
...                     ...       ...       ...       ...       ...  
2026-08-31 XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249  
           YUM     4.974626  5.014113  5.053600  3.248934 -1.234277  
           ZBH     4.485455  4.555733  4.626010  0.354879  0.997090  
           ZBRA    5.402393  5.721512  6.040631  1.556197  3.605673  
           ZTS     4.303346  4.340344  4.377342 -0.398774 -0.195375  

[48791 rows x 9 columns]

In [11]:
data = pd.concat([df.unstack('ticker')['dollar_volume'].resample('ME').mean().stack('ticker').to_frame('dollar_volume'),
                  df.unstack()[last_cols].resample('ME').last().stack('ticker')],axis=1)

Axis =1 concatinate kiye hai kyuki dono table side by side jodna hai 

In [12]:
data

dollar_volume       close        high  garman_klass_vol  \
date       ticker                                                            
2018-08-31 A          153.014041   63.639286   63.648711          0.000014   
           AAPL      5854.973599   53.774918   54.067851          0.000070   
           ABBV       294.747135   68.465599   68.972064          0.000048   
           ABNB              NaN         NaN         NaN               NaN   
           ABT        248.662236   57.822296   58.168332          0.000046   
...                          ...         ...         ...               ...   
2026-08-31 XYZ        476.457452   78.279999   79.199997          0.000312   
           YUM        370.952232  150.339996  150.389999          0.000234   
           ZBH        306.447240   97.889999   97.940002          0.000526   
           ZBRA       389.639119  378.369995  386.230011          0.000425   
           ZTS        699.134300   73.540001   74.389999          0.000487   

                         rsi    bb_low    bb_mid   bb_high       atr      macd  
date       ticker                                                               
2018-08-31 A       94.747320       NaN       NaN       NaN       NaN       NaN  
           AAPL    96.803688       NaN       NaN       NaN       NaN       NaN  
           ABBV    75.028397       NaN       NaN       NaN       NaN       NaN  
           ABNB          NaN       NaN       NaN       NaN       NaN       NaN  
           ABT     96.406869       NaN       NaN       NaN       NaN       NaN  
...                      ...       ...       ...       ...       ...       ...  
2026-08-31 XYZ     49.935831  4.342263  4.398908  4.455553 -0.668036  0.144249  
           YUM     47.965239  4.974626  5.014113  5.053600  3.248934 -1.234277  
           ZBH     59.070244  4.485455  4.555733  4.626010  0.354879  0.997090  
           ZBRA    78.001656  5.402393  5.721512  6.040631  1.556197  3.605673  
           ZTS     43.338929  4.303346  4.340344  4.377342 -0.398774 -0.195375  

[48791 rows x 10 columns]

In [13]:
data = data.dropna()

In [14]:
data

dollar_volume       close        high  garman_klass_vol  \
date       ticker                                                            
2018-09-30 A          123.257081   66.466026   66.899459          0.000027   
           AAPL      7483.296675   53.328423   53.352044          0.000026   
           ABBV       346.325562   67.466965   67.880699          0.000083   
           ABT        344.906256   63.462650   63.652970          0.000037   
           ACGL        33.195092   28.346285   28.393831          0.000063   
...                          ...         ...         ...               ...   
2026-08-31 XYZ        476.457452   78.279999   79.199997          0.000312   
           YUM        370.952232  150.339996  150.389999          0.000234   
           ZBH        306.447240   97.889999   97.940002          0.000526   
           ZBRA       389.639119  378.369995  386.230011          0.000425   
           ZTS        699.134300   73.540001   74.389999          0.000487   

                         rsi    bb_low    bb_mid   bb_high       atr      macd  
date       ticker                                                               
2018-09-30 A       78.358819  4.149613  4.189401  4.229189 -2.169555  0.494001  
           AAPL    69.446139  3.949469  3.980092  4.010716 -1.444815 -0.224777  
           ABBV    57.694161  4.192278  4.220161  4.248044 -1.099658 -0.603603  
           ABT     89.293170  4.030530  4.101133  4.171736 -2.043854  1.063747  
           ACGL    55.409436  3.369646  3.389962  3.410277 -1.636060 -0.501939  
...                      ...       ...       ...       ...       ...       ...  
2026-08-31 XYZ     49.935831  4.342263  4.398908  4.455553 -0.668036  0.144249  
           YUM     47.965239  4.974626  5.014113  5.053600  3.248934 -1.234277  
           ZBH     59.070244  4.485455  4.555733  4.626010  0.354879  0.997090  
           ZBRA    78.001656  5.402393  5.721512  6.040631  1.556197  3.605673  
           ZTS     43.338929  4.303346  4.340344  4.377342 -0.398774 -0.195375  

[47211 rows x 10 columns]

In [15]:
data['dollar_volume'].unstack('ticker')

ticker,A,AAPL,ABBV,ABNB,ABT,ACGL,ACN,ADBE,ADI,ADM,...,WY,WYNN,XEL,XOM,XYL,XYZ,YUM,ZBH,ZBRA,ZTS
date,,,,,,,,,,,,,,,,,,,,,
2018-09-30,123.257081,7483.296675,346.325562,NaN,344.906256,33.195092,322.641604,849.769324,272.947930,119.569277,...,91.033384,359.163483,133.277889,700.481940,64.043085,1241.184916,125.538310,125.658699,55.846486,171.697576
2018-10-31,176.723221,7139.701668,375.788405,NaN,478.142171,27.136687,332.837455,1163.944148,299.348712,134.238429,...,115.469611,282.922345,132.865101,760.328145,80.912903,1582.894439,153.519280,138.478795,77.547178,199.400395
2018-11-30,187.605549,8333.096335,485.583011,NaN,392.594305,28.335316,324.441680,920.189366,253.254916,141.734281,...,86.996433,355.582169,217.376932,765.462942,68.818237,1223.181116,178.192640,115.918710,84.510682,208.652350
2018-12-31,172.944904,7319.299544,418.578258,NaN,463.053114,32.324502,408.404716,1167.567529,213.441554,144.001852,...,138.181374,314.208579,213.792334,1029.906300,66.983520,1096.528512,215.135557,161.692742,80.346548,185.721949
2019-01-31,138.812096,5738.320202,528.807148,NaN,456.949014,24.491152,270.300775,772.400237,232.806527,107.001201,...,99.516470,271.418221,155.913080,650.139993,63.267197,1282.598422,156.325614,126.048645,80.104116,201.561326
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-04-30,212.683888,11433.614858,1394.386418,438.839102,1190.327935,200.588254,955.879566,1273.139998,1169.567012,227.818027,...,106.186182,139.359100,357.710538,3045.493481,313.615902,333.124385,250.701627,229.012970,175.222064,380.488173
2026-05-31,312.257707,14562.781103,1158.987803,563.194936,1080.778751,246.003800,1029.567463,1170.459676,1864.762579,332.057695,...,129.280464,156.121431,448.021813,2649.690346,220.429735,452.449377,287.640510,188.879403,308.030033,799.470692
2026-06-30,313.490334,19562.997080,2035.292437,557.963874,1252.316612,217.615333,1539.782603,1892.044633,2340.480108,284.722074,...,146.661560,160.479344,505.597671,2577.383586,220.503832,463.372034,358.197746,198.361409,228.071479,572.913364


In [16]:
data['dollar_volume']=(data.loc[:, 'dollar_volume'].unstack('ticker').rolling(5*12, min_periods=12).mean().stack())

In [17]:
data

dollar_volume       close        high  garman_klass_vol  \
date       ticker                                                            
2018-09-30 A                 NaN   66.466026   66.899459          0.000027   
           AAPL              NaN   53.328423   53.352044          0.000026   
           ABBV              NaN   67.466965   67.880699          0.000083   
           ABT               NaN   63.462650   63.652970          0.000037   
           ACGL              NaN   28.346285   28.393831          0.000063   
...                          ...         ...         ...               ...   
2026-08-31 XYZ        868.471911   78.279999   79.199997          0.000312   
           YUM        243.813358  150.339996  150.389999          0.000234   
           ZBH        183.563650   97.889999   97.940002          0.000526   
           ZBRA       150.770095  378.369995  386.230011          0.000425   
           ZTS        418.616393   73.540001   74.389999          0.000487   

                         rsi    bb_low    bb_mid   bb_high       atr      macd  
date       ticker                                                               
2018-09-30 A       78.358819  4.149613  4.189401  4.229189 -2.169555  0.494001  
           AAPL    69.446139  3.949469  3.980092  4.010716 -1.444815 -0.224777  
           ABBV    57.694161  4.192278  4.220161  4.248044 -1.099658 -0.603603  
           ABT     89.293170  4.030530  4.101133  4.171736 -2.043854  1.063747  
           ACGL    55.409436  3.369646  3.389962  3.410277 -1.636060 -0.501939  
...                      ...       ...       ...       ...       ...       ...  
2026-08-31 XYZ     49.935831  4.342263  4.398908  4.455553 -0.668036  0.144249  
           YUM     47.965239  4.974626  5.014113  5.053600  3.248934 -1.234277  
           ZBH     59.070244  4.485455  4.555733  4.626010  0.354879  0.997090  
           ZBRA    78.001656  5.402393  5.721512  6.040631  1.556197  3.605673  
           ZTS     43.338929  4.303346  4.340344  4.377342 -0.398774 -0.195375  

[47211 rows x 10 columns]

In [18]:
data['dollar_vol_rank'] = data.groupby('date')['dollar_volume'].rank(ascending=False)

In [19]:
data

dollar_volume       close        high  garman_klass_vol  \
date       ticker                                                            
2018-09-30 A                 NaN   66.466026   66.899459          0.000027   
           AAPL              NaN   53.328423   53.352044          0.000026   
           ABBV              NaN   67.466965   67.880699          0.000083   
           ABT               NaN   63.462650   63.652970          0.000037   
           ACGL              NaN   28.346285   28.393831          0.000063   
...                          ...         ...         ...               ...   
2026-08-31 XYZ        868.471911   78.279999   79.199997          0.000312   
           YUM        243.813358  150.339996  150.389999          0.000234   
           ZBH        183.563650   97.889999   97.940002          0.000526   
           ZBRA       150.770095  378.369995  386.230011          0.000425   
           ZTS        418.616393   73.540001   74.389999          0.000487   

                         rsi    bb_low    bb_mid   bb_high       atr  \
date       ticker                                                      
2018-09-30 A       78.358819  4.149613  4.189401  4.229189 -2.169555   
           AAPL    69.446139  3.949469  3.980092  4.010716 -1.444815   
           ABBV    57.694161  4.192278  4.220161  4.248044 -1.099658   
           ABT     89.293170  4.030530  4.101133  4.171736 -2.043854   
           ACGL    55.409436  3.369646  3.389962  3.410277 -1.636060   
...                      ...       ...       ...       ...       ...   
2026-08-31 XYZ     49.935831  4.342263  4.398908  4.455553 -0.668036   
           YUM     47.965239  4.974626  5.014113  5.053600  3.248934   
           ZBH     59.070244  4.485455  4.555733  4.626010  0.354879   
           ZBRA    78.001656  5.402393  5.721512  6.040631  1.556197   
           ZTS     43.338929  4.303346  4.340344  4.377342 -0.398774   

                       macd  dollar_vol_rank  
date       ticker                             
2018-09-30 A       0.494001              NaN  
           AAPL   -0.224777              NaN  
           ABBV   -0.603603              NaN  
           ABT     1.063747              NaN  
           ACGL   -0.501939              NaN  
...                     ...              ...  
2026-08-31 XYZ     0.144249             66.0  
           YUM    -1.234277            291.0  
           ZBH     0.997090            362.0  
           ZBRA    3.605673            399.0  
           ZTS    -0.195375            169.0  

[47211 rows x 11 columns]

Ye pura block kya kar raha hai, step by step:

Columns ko alag kiya — do hisso mein baanta: dollar_volume (jise alag tareeke se handle karna hai) aur baaki saare indicators (rsi, macd, atr, bb_low/mid/high, garman_klass_vol, high) — raw noise wale columns (volume, open, low) ko chhod diya.
Daily se Monthly banaya — daily data ko monthly mein convert kiya, par do alag-alag tareeko se:
dollar_volume → mahine ka average (mean) liya, kyunki liquidity ko poore mahine ka typical behavior dikhana chahiye, sirf ek din ka nahi.
Baaki indicators → mahine ke aakhri din (last) ki value li, kyunki ye sab "abhi ki sthiti" (current snapshot) batate hain — inhe average karne se matlab kho jata.
NaN values hataayi — jo rows abhi bhi missing data wali thi (indicators ke warm-up period ki wajah se, ya jo stocks abhi exist hi nahi karte the jaise IPO se pehle).
dollar_volume ko stabilize kiya — dollar_volume ko uske apne 5-saal (60 mahine) ke rolling average se overwrite kar diya, taaki ek-aadh mahine ka achaanak spike poori picture na bigade — sirf lagatar/sustained liquidity hi mayne rakhe. Isse shuruaati 5 saal ki rows phir se NaN ho gayi (kyunki itna history nahi tha rolling window bharne ke liye).
Har mahine alag se rank kiya — groupby('date') ka istemal karke, har mahine ke andar stocks ko unke (ab stabilize hue) dollar_volume ke hisaab se rank kiya — sabse zyada liquid ko rank 1 mila.
Sirf top 150 rakhe — har mahine ke sabse zyada liquid ~149 stocks hi rakhe — ye koi fixed/permanent list nahi hai, balki har mahine badalne wali, dynamic list hai.
Scaffolding columns hataaye — dollar_volume aur dollar_vol_rank ko hata diya, kyunki unka kaam (filtering ke liye) ho chuka tha — ye ML model ke liye actual useful/predictive features nahi hain.

Nateeja: Ab dataset monthly hai, har mahine sirf sabse zyada consistently liquid ~149 stocks par focused hai, aur har row mein us mahine ki latest technical-indicator readings maujood hain — agla step (returns calculate karna, phir ML model mein feed karna) karne ke liye ready hai.

In [20]:
data = data[data['dollar_vol_rank']<150].drop(['dollar_volume','dollar_vol_rank'],axis=1)

In [21]:
data

close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2019-08-31 AAPL     50.077595   50.487829          0.000103  54.159012   
           ABBV     49.354843   49.872868          0.000063  44.137036   
           ABT      75.040871   75.304730          0.000066  52.158314   
           ACN     177.113998  179.357300          0.000161  57.729453   
           ADBE    284.510010  288.500000          0.000395  45.672844   
...                       ...         ...               ...        ...   
2026-08-31 WDC     454.100006  467.799988          0.000662  43.796277   
           WFC      88.930000   89.220001          0.000170  59.174628   
           WMT     116.010002  116.050003          0.000153  54.108979   
           XOM     158.722244  159.199152          0.000119  62.142821   
           XYZ      78.279999   79.199997          0.000312  49.935831   

                     bb_low    bb_mid   bb_high       atr      macd  
date       ticker                                                    
2019-08-31 AAPL    3.861528  3.913774  3.966021 -1.351044 -0.200506  
           ABBV    3.879913  3.915435  3.950958 -1.157990 -0.423443  
           ABT     4.295560  4.322496  4.349432 -1.089955 -0.268489  
           ACN     5.117288  5.156513  5.195739 -1.185125  0.279742  
           ADBE    5.634026  5.664694  5.695362 -1.113644 -0.295100  
...                     ...       ...       ...       ...       ...  
2026-08-31 WDC     6.028525  6.205588  6.382652  5.179602 -4.059220  
           WFC     4.444806  4.474222  4.503639  1.216887  0.744505  
           WMT     4.693527  4.727599  4.761672  2.212910 -0.599633  
           XOM     4.988154  5.038804  5.089454  2.227161  1.943697  
           XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249  

[12665 rows x 9 columns]

In [22]:
a = df.xs('AAPL',level=1)

In [23]:
a

Price,close,high,low,open,volume,garman_klass_vol,rsi,bb_low,bb_mid,bb_high,atr,macd,dollar_volume
date,,,,,,,,,,,,,
2018-08-15,49.666737,49.784856,49.215521,49.425773,115230400.0,0.000057,NaN,NaN,NaN,NaN,NaN,NaN,5723.117925
2018-08-16,50.394360,50.510114,49.957317,50.023464,114001600.0,0.000039,100.000000,NaN,NaN,NaN,NaN,NaN,5745.037624
2018-08-17,51.400726,51.488133,50.356554,50.422700,141708000.0,0.000104,100.000000,NaN,NaN,NaN,NaN,NaN,7283.894125
2018-08-20,50.899899,51.778702,50.817214,51.523567,121150800.0,0.000118,96.567432,NaN,NaN,NaN,NaN,NaN,6166.563427
2018-08-21,50.800686,51.308600,50.562086,51.216468,104639200.0,0.000082,95.881143,NaN,NaN,NaN,NaN,NaN,5315.743130
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-08-06,312.140808,316.017470,308.963556,314.069138,46139900.0,0.000240,49.622970,5.711609,5.780952,5.850296,3.186242,0.149707,14402.145672
2026-08-07,313.059998,314.538733,310.472233,311.181643,34437200.0,0.000071,50.103315,5.710758,5.780637,5.850515,2.977069,-0.016988,10780.909748
2026-08-10,308.260010,308.260010,304.609985,306.829987,44812500.0,0.000063,47.608072,5.706730,5.779237,5.851745,2.957197,-0.278731,13813.901688


Theek hai, sirf .xs() wali cheez ka summary:

df.xs('AAPL', level=1) — Summary
python
a = df.xs('AAPL', level=1)
.xs() (cross-section) ka kaam hai MultiIndex se ek specific value ki saari rows nikalna, bina poore DataFrame ko group/reshape kiye
'AAPL' — jo value dhundni hai
level=1 — batata hai ki 'AAPL' ko index ke kis level mein dhundo (tumhare df mein level=0 = date, level=1 = ticker)

Result: a mein sirf AAPL ki saari rows aayengi (har date ke liye), aur chunki humne ticker pe filter kiya, ticker level index se hat jaata hai (redundant ho gaya tha, kyunki ab wo hamesha AAPL hi hoga) — sirf date index ke roop mein reh jaata hai:

             close     high     rsi    macd   ...
date
2018-08-15   49.67     49.78    NaN    NaN
2018-08-16   50.10     50.30    NaN    NaN
...

Kyun useful hai: Ab tak jo bhi groupby(level=1) se kiya, wo saare ~500 stocks pe ek saath chalta tha. Lekin kabhi-kabhi debugging ya analysis ke liye sirf ek specific stock ko alag se check karna hota hai (jaise "AAPL ka RSI, MACD kaisa dikh raha hai") — .xs() ye kaam ek line mein aasaan bana deta hai.

In [24]:
def calculate_returns(df):

    outlier_cutoff = 0.005

    lags = [1, 2, 3, 6, 9, 12]

    for lag in lags:

        df[f'return_{lag}m'] = (df['close']
                              .pct_change(lag)
                              .pipe(lambda x: x.clip(lower=x.quantile(outlier_cutoff),
                                                     upper=x.quantile(1-outlier_cutoff)))
                              .add(1)
                              .pow(1/lag)
                              .sub(1))
    return df

In [25]:
data = data.groupby(level=1, group_keys=False).apply(calculate_returns).dropna()

In [26]:
data

close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2020-08-31 AAPL    125.057526  126.957042          0.000707  76.744809   
           ABBV     75.952614   76.142956          0.000105  52.776389   
           ABT      97.865204   99.894561          0.000204  67.784280   
           ACN     217.961929  221.241397          0.000094  67.876178   
           ADBE    513.390015  516.729980          0.000153  66.121887   
...                       ...         ...               ...        ...   
2026-08-31 WDC     454.100006  467.799988          0.000662  43.796277   
           WFC      88.930000   89.220001          0.000170  59.174628   
           WMT     116.010002  116.050003          0.000153  54.108979   
           XOM     158.722244  159.199152          0.000119  62.142821   
           XYZ      78.279999   79.199997          0.000312  49.935831   

                     bb_low    bb_mid   bb_high       atr      macd  \
date       ticker                                                     
2020-08-31 AAPL    4.634849  4.743995  4.853141 -0.215957  1.968695   
           ABBV    4.304350  4.329241  4.354133 -1.060942 -0.352670   
           ABT     4.460974  4.528430  4.595886  0.635079  1.529678   
           ACN     5.324580  5.364798  5.405016 -1.217038  0.870915   
           ADBE    6.036542  6.152162  6.267782  0.679071  1.881340   
...                     ...       ...       ...       ...       ...   
2026-08-31 WDC     6.028525  6.205588  6.382652  5.179602 -4.059220   
           WFC     4.444806  4.474222  4.503639  1.216887  0.744505   
           WMT     4.693527  4.727599  4.761672  2.212910 -0.599633   
           XOM     4.988154  5.038804  5.089454  2.227161  1.943697   
           XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249   

                   return_1m  return_2m  return_3m  return_6m  return_9m  \
date       ticker                                                          
2020-08-31 AAPL     0.204976   0.176611   0.158586   0.110026   0.076700   
           ABBV     0.009061  -0.006341   0.015125   0.023252   0.014328   
           ABT      0.087738   0.096338   0.050042   0.059706   0.029314   
           ACN      0.067399   0.059010   0.060994   0.049919   0.021139   
           ADBE     0.155451   0.085986   0.099162   0.068430   0.057830   
...                      ...        ...        ...        ...        ...   
2026-08-31 WDC     -0.166544  -0.156820  -0.050854   0.084251   0.281101   
           WFC      0.034593   0.040332   0.048744   0.016685   0.005747   
           WMT      0.043255   0.012067   0.000748  -0.015546   0.006081   
           XOM      0.027728   0.080947   0.032203   0.008908   0.037822   
           XYZ     -0.036435   0.014889   0.011145   0.034948   0.017777   

                   return_12m  
date       ticker              
2020-08-31 AAPL      0.073806  
           ABBV      0.036576  
           ABT       0.022377  
           ACN       0.017444  
           ADBE      0.050419  
...                       ...  
2026-08-31 WDC       0.195755  
           WFC       0.008411  
           WMT       0.015550  
           XOM       0.030196  
           XYZ      -0.001434  

[10419 rows x 15 columns]

Ye function har stock ke liye 6 alag-alag time horizons (1, 2, 3, 6, 9, 12 mahine) ke returns calculate karta hai — momentum/trend features banane ke liye.

Step-by-Step Breakdown
1. pct_change(lag)
lag mahine pehle ke price se abhi ke price ka percentage change nikalta hai
Example: agar 3 mahine pehle price $100 tha, ab $115 hai → pct_change(3) = 0.15 (15%)
2. .pipe(lambda x: x.clip(...)) — Outlier Clipping
.pipe() — method chain ke beech mein custom logic daalne ka tareeka, jab data ko khud reference karna ho
.pipe() ke bina dikkat: chain ke beech mein result ka koi naam nahi hota, isliye x.quantile(...) jaisa kuch likhna mushkil hota
.pipe(lambda x: ...) us in-between result ko x naam de deta hai, jisse use do baar reference kar sako
x.quantile(0.005) — sabse neeche wale 0.5% extreme (bahut negative) returns ki value
x.quantile(0.995) — sabse upar wale 0.5% extreme (bahut positive) returns ki value
.clip(lower=..., upper=...) — jo values in extremes se bahar hain, unhe cap kar deta hai (guard rails jaisa) — bina un rows ko delete kiye, sirf unhe "reasonable extreme" tak seemit kar diya
Kyun zaroori hai: kabhi-kabhi stock returns mein freak/glitch values aa jaati hain (jaise -90% ya +500%), jo ML model ko confuse kar sakti hain — clipping unhe control mein rakhta hai
3. .add(1).pow(1/lag).sub(1) — Per-Month Equivalent Rate mein Convert Karna

Problem: 3-mahine ka 15% return aur 1-mahine ka 15% return — dono bilkul alag cheezein hain (ek dheere-dheere hua, doosra bahut tezi se). Inhe directly compare nahi kar sakte.

Solution — teen steps:

.add(1) → return ko "growth factor" banaya (0.15 → 1.15, matlab paisa 1.15 guna hua)
.pow(1/lag) → growth factor ko lag barabar hisso mein todkar monthly rate nikali (jaise 3-mahine ka total factor 1.15 hai, to cube root nikal ke pata chalta hai har mahine kitna bada, jo 3 baar compound hoke 1.15 banega)
.sub(1) → wapas plain percentage return mein badla (growth factor se 1 ghata diya)

Verify kiya numbers se:

0.15 → add(1) → 1.15 → pow(1/3) → 1.0477 → sub(1) → 0.0477 (≈4.77% per month)
Check: 1.0477 × 1.0477 × 1.0477 ≈ 1.15 ✓

Sanity check (lag=1): agar lag=1 ho, to formula kuch nahi badalta — jo already sahi hai, kyunki 1-mahine ka return already "per month" hai.

4. Special Case — Agar Stock Constant Rate se Grow Kare

Agar koi stock har mahine exactly same rate (jaise 5%) se badhe, to return_1m, return_2m, return_3m — sab same (5%) aayenge. Ye verify kiya tha actual numbers se:

Month 0: 100 → Month 1: 105 → Month 2: 110.25 → Month 3: 115.76
return_1m = 5%, return_2m = 5%, return_3m = 5%  (sab same)

Kyun hota hai: kyunki .add(1).pow(1/lag).sub(1) formula ka pura maksad hi ye hai ki har return ko "per-month equivalent rate" mein convert kare — agar growth khud constant rate se ho rahi hai, to uska per-month equivalent bhi wahi constant rate hi niklega, chahe tum use 1, 2, ya 12 mahine ke window se naapo.

Real world mein: actual stocks kabhi bhi bilkul constant rate se nahi badhte, isliye real data mein ye columns alag-alag hi aayenge — aur yahi difference ML model ke liye useful signal hai:

Agar return_1m zyada hai but return_12m kam → recent momentum hai
Agar return_12m zyada hai but return_1m kam/negative → pura saal accha chala tha but abhi dheema ho raha hai
Agar sab lags lagbhag barabar hain → steady, consistent growth
5. groupby(level=1, group_keys=False).apply(calculate_returns)
groupby(level=1) → har stock ke liye alag se returns calculate ho, kisi doosre stock ka data mix na ho
.apply() (na ki .transform()) → kyunki function ek saath 6 naye columns bana raha hai (poora sub-DataFrame chahiye, ek column nahi)
group_keys=False → result mein extra/duplicate ticker-label index add hone se rokta hai
6. .dropna()
Jo rows abhi bhi missing hain unhe hata deta hai — mainly early months jinke liye poora lag period (jaise 12-mahine ka data) available nahi tha
Ek Error Bhi Discuss Hua Tha

KeyError: 'close' — kyunki pichle steps mein close column ko last_cols se exclude kar diya gaya tha, isliye jab tak data bana, close column already gayab tha. Fix: last_cols wali exclusion list se close ko hataya, taaki wo aage bhi data mein bana rahe.

In [28]:
data.to_pickle('data/03_top_liquid_stocks.pkl')